### PR TITLE
fix: Indent overhaul

### PR DESCRIPTION
**Pull request by bricoletc**
Not yet ready to merge this but want to log what passes and what doesn't

#159 solved
#161 solved
#160: first part not solved (comments still get dedented), second part is solved (comments don't magically disappear anymore..!!poof!!..)
#157: the example raised by @weber8thomas and the first one by @corneliusroemer solved. However @corneliusroemer raised a good test here:
```
    def test_if_statement_with_snakecode_2comments_snakecode_inside(self):
        """https://github.com/snakemake/snakefmt/issues/159"""
        snakecode = (
            "if True:\n\n"
            f"{TAB * 1}ruleorder: a > b\n"
            "\n"
            f"# comment\n"
            f"# comment\n"
            f"{TAB * 1}ruleorder: c > d\n"
        )
        formatter = setup_formatter(snakecode)
        assert formatter.get_formatted() == snakecode
```

Which does not pass. I will add it to this PR and try to fix it.

### CODE REVIEW

**Comment by bricoletc** [comment]
It's fine to review or not now @mbhall88  (it's always good to review of course - but i will add more code to this PR)

**Review by mbhall88 — APPROVED** [review]
Looks great so far. Thanks for this!

**Comment by bricoletc** [comment]
@mbhall this PR 'solves' the four issues above.
I put 'solves' in quotes because comment parsing and formatting is really tricky. 
The `tokenizer` module from python stdlib does not output INDENT/DEDENT tokens when indents/dedents precede comments - instead, it outputs comments first. This is v. hard to deal with.

But AFAICT all four issues mentioned above are resolved.

**Comment by bricoletc** [comment]
If you or the people in the four issues can double-check things now fine (e.g. paste their code blocks into a .smk file, run snakefmt on them) , would be good - but I'm happy enough to say we can merge this, bump to 0.8.0, and close these issues

File: tests/test_formatter.py [hunk 1]
```diff
@@ -964,6 +970,30 @@ def test_no_inline_comments_stay_untouched(self):
         formatter = setup_formatter(snakecode)
         assert formatter.get_formatted() == snakecode
 
+    def test_snakecode_after_indented_comment_does_not_get_unindented(self):
+        """https://github.com/snakemake/snakefmt/issues/159#issue-1441174995"""
+        snakecode = (
+            'if config.get("s3_dst"):\n\n'
+            f'{TAB * 1}include: "workflow/rule1.smk"\n'
+            f'{TAB * 1}include: "workflow/rule2.smk"\n'
+            f"{TAB * 1}# a comment\n"
```
Comment (mbhall88) [inline tests/test_formatter.py:979]: Shouldn't there be a new line between these two lines?

**Review by mbhall88 — COMMENTED** [review]
Amazing work Brice! When testing out on the example in #160 I noticed a small thing. I think there should be a new line between the last include before the first comment?

File: tests/test_formatter.py (hunk 1 above)
Comment (bricoletc) [reply tests/test_formatter.py:979]: I don't think so. our policy has been that if you have a series of consecutive keywords (here, `include`s), *possibly interspersed with comments* you don't space any of the lines out

Comment (bricoletc) [reply tests/test_formatter.py:979]: If you replace the final `include` with, say, `shell`, you will get that newline before the first comment

Comment (mbhall88) [reply tests/test_formatter.py:979]: Ahhh very nice!! Too smart for me

**Review by mbhall88 — APPROVED** [review]
This is great. Good to merge when you're ready

**Comment by bricoletc** [comment]
This is now the tip of master; weirdly it doesn't appear as merged

**Comment by bricoletc** [comment]
Thought you meant I should merge this - feel free to change/edit as you want - also lets do 0.8.0 ?

**Pull request merged**